Add DFRobot CAN bus Arduino UNO R3 shield

nixward opened the pull request:
Adds the DFRobot Arduino Uno R3 compatible CAN bus v2.0 shield based on the MCP2515 CAN controller.

- updates CAN sample to use common Zephyr CAN bus-speed of 125kbit/s.
- updates CAN sample documentation for building for integrated and stand alone (MCP2515) CAN controllers, adding an alternate conf file to build with MCP2515 driver without warning.
- additionally configures Arduino Uno R3 connector SPI peripheral by default if general SPI is configured for the nRF52_pca10040 and nRF52840_pca_10056 boards.

nixward committed:
boards: shields: Add DFRobot CAN bus Shield V2.0

Useful for prototyping designs for the MCP2515 CAN controller

Signed-off-by: Nick Ward <nix.ward@gmail.com>

karstenkoenig reviewed (commented):
Cool! Looks very good overall, and with nice documentation too!

karstenkoenig commented on drivers/can/Kconfig.mcp2515:
```diff
@@ -34,7 +34,7 @@ config CAN_MCP2515_INT_THREAD_PRIO
 	  interrupt handling and incoming packets.
 
-config CAN_MCP2515_MAX_FILTER
+config CAN_MAX_FILTER
```
Oh right, the mcux and stm32 Kconfig have a ```depends on``` on the symbol. Thanks for sorting this out :-)

alexanderwachter replied:
The problem here is that the MCP2515 could be used as an additional CAN controller beside the internal. Then we would run into problems here!

nixward replied:
Correct me if I'm wrong but wouldn't using the MCP2515 as a 2nd CAN controller be exceptional?  I imagine the majority of users of the MCP2515 would be adding a CAN controller to a SOC that didn't have one.  With this change those users can build the CAN sample without a Kconfig warning regarding CAN_MAX_FILTER.

alexanderwachter replied:
True. Let's use CAN_MAX_FILTER for all drivers.

nixward replied:
Apologies I meant to send this earlier, I've been thinking about this and I've come around to agree with you that it would be nice to make it straightforward to use it as a 2nd CAN controller (I missed this intent when I first saw the Kconfig warning and was trying to understand the code).  It dawned on me yesterday there's another way to get rid of the compile warning with an extra conf file, it will require a bit more documentation, I'll try to push the changes tomorrow.

alexanderwachter replied:
The good thing about only using CAN_MAX_FILTER is that there is only once config for all boards and samples. It makes it easy to migrate apps.

nixward replied:
An alternate CAN sample conf file is now added for building the sample using the MCP2515 driver.

dbkinder commented on boards/shields/dfrobot_can_bus_v2_0/boards/nrf52840_pca10056.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_SPI_3=y
+CONFIG_SPI_3_NRF_SPIM=y
```
Looks like some of your files are missing the final newline...

nixward replied:
Thanks, now fixed.

dbkinder reviewed (approved):
docs LGTM, thanks!

erwango commented on boards/shields/dfrobot_can_bus_v2_0/dfrobot_can_bus_v2_0.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_GPIO=y
+CONFIG_SPI=y
+CONFIG_CAN=y
```
Should not be required as this is would be done on application side.
An application should enable CAN, either CAN controller this is a SoC IP or a device mounted on board or shield.
(And in short term, for application portability, other flags should be under if CAN condition, but we miss a piece in shield support to do that yet).

nixward replied:
I see, thanks for that, done

nixward replied:
@erwango is this issue now resolved?

alexanderwachter commented on boards/shields/dfrobot_can_bus_v2_0/dfrobot_can_bus_v2_0.overlay:
```diff
@@ -0,0 +1,26 @@
+		status = "okay";
+		label = "CAN_1";
+		reg = <0x0>;
+		bus-speed = <250000>;
```
Please use 125000 here. We agreed to have a common speed for all boards.

nixward replied:
No problem.  For my future reference would you have a link for this agreement and any others?  Thanks

nixward replied:
@alexanderwachter  I've updated the bus-speed to 125000 and (if all ok) added two commits to fix this issue in other MCP2515 DTS files and one spot in the CAN driver sample which was also using 250000.

alexanderwachter replied:
Thanks for fixing the sample too. This "agreement" was made after I noticed that the boards are not compatible with each other. It's not written down somewhere.

nixward committed:
samples: drivers: CAN: Use common Zephyr bus-speed of 125kbit/s

Consistent use of 125000bit/s for Zephyr CAN code

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward commented on boards/shields/link_board_can/link_board_can.overlay:
```diff
@@ -13,13 +13,12 @@
 
 	mcp2515@0 {
 		compatible = "microchip,mcp2515";
-		spi-port-name = "SPI_1";
```
I forgot to add this to my initial PR but after becoming aware that the 125kbit/s bus-speed was the common default value I thought it was a good time to add these two changes.  I had previously discussed with @karstenkoenig and he'd agreed that the spi-port-name property was invalid and should be removed.

nixward replied:
This has been fixed in PR #19680

erwango commented on boards/shields/dfrobot_can_bus_v2_0/dfrobot_can_bus_v2_0.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_GPIO=y
```
GPIO shouldn't be needed, as it's part of default board configuration (ie boards should have it enabled by default)

nixward replied:
Thanks, it has been removed now

erwango reviewed (changes requested):
Last comment otherwise LGTM

nixward committed:
samples: drivers: CAN: Update to build cleanly for MCP2515 driver

Add alternate conf file for MCP2515 so it builds without warning.
Update documentation.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

alexanderwachter commented:
@erwango can you give your +1 now?

jfischer-phytec-iot commented:
@alexanderwachter @karstenkoenig As this shield can also be used in generic way could we remove mcp2515-dts.overlay from zephyr/samples/drivers/CAN ?

alexanderwachter commented:
> @alexanderwachter @karstenkoenig As this shield can also be used in generic way could we remove mcp2515-dts.overlay from zephyr/samples/drivers/CAN ?

Sure.

jfischer-phytec-iot commented on boards/shields/dfrobot_can_bus_v2_0/boards/nrf52840_pca10056.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_SPI_3=y
```
@erwango This above is not consistent.

nixward replied:
@jfischer-phytec-iot this is configuration of the nrf52840_pca10056 board for use with the DFRobot CAN bus shield.

jfischer-phytec-iot replied:
I know what it is and I had to remove it for the shields I recently added.

alexanderwachter replied:
@jfischer-phytec-iot SPI_3 is not enabled by default on this board. We maybe want to enable it, but I would suggest doing this in a separate PR. So far, we have to enable it here.

nixward replied:
@jfischer-phytec-iot I found the PR issue you had, you had it right, I've added a https://github.com/zephyrproject-rtos/zephyr/pull/19124#discussion_r333437641

nixward replied:
@jfischer-phytec-iot thanks for making me aware of this issue, the files are now removed and required board changes made

karstenkoenig commented:
> @alexanderwachter @karstenkoenig As this shield can also be used in generic way could we remove mcp2515-dts.overlay from zephyr/samples/drivers/CAN ?

Agreed, @nixward your shield covers everything that overlay does and it's properly done as a shield, please remove that file

nixward committed:
samples: drivers: CAN: Remove legacy MCP2515 overlay

DFRobot CAN bus or link_board_can shields can now be used.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward commented:
> > @alexanderwachter @karstenkoenig As this shield can also be used in generic way could we remove mcp2515-dts.overlay from zephyr/samples/drivers/CAN ?
> 
> Agreed, @nixward your shield covers everything that overlay does and it's properly done as a shield, please remove that file.
> Besides that we should just get this merged, this is a great shield and documents really well how to get started quickly

Cheers! and mcp2515-dts.overlay now removed

nixward commented:
As per @jfischer-phytec-iot review, [for software compatibility, boards must also provide a configuration matching their supported connectors](https://docs.zephyrproject.org/latest/guides/porting/shields.html?highlight=shield#board-compatibility) I've enabled the SPI peripherals for the nrf52832 and nrf52840 DK boards I was testing the Arduino shield with.

nixward committed:
boards: nrf52_pca10040:  When SPI is required enable SPI_2 by default

Required by Arduino Uno Rev3 shield connector

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward committed:
boards: nrf52840_pca10056:  When SPI is required enable SPI_3 by default

Required by Arduino Uno Rev3 shield connector

Signed-off-by: Nick Ward <nix.ward@gmail.com>

carlescufi commented:
@jfischer-phytec-iot would you take another look please?
